pmd-java/src/main/java/net/sourceforge/pmd/lang/java/metrics/impl/internal/CognitiveComplexityVisitor.java

```diff
@@ -0,0 +1,307 @@
+        State state = (State) data;
+
+        state.structuralComplexity();
+        super.visit(node, data);
+        state.decreaseNestingLevel();
+
+        return data;
+    }
+
+    @Override
+    public Object visit(ASTCatchStatement node, Object data) {
+        State state = (State) data;
+
+        state.structuralComplexity();
+        super.visit(node, data);
+        state.decreaseNestingLevel();
+
+        return data;
+    }
+
+    @Override
+    public Object visit(ASTDoStatement node, Object data) {
+        State state = (State) data;
+
+        state.structuralComplexity();
+        super.visit(node, data);
+        state.decreaseNestingLevel();
+
+        return data;
+    }
+
+    @Override
+    public Object visit(ASTConditionalExpression node, Object data) {
+        State state = (State) data;
+
+        state.structuralComplexity();
+        super.visit(node, data);
+        state.decreaseNestingLevel();
+
+        return data;
+    }
+
+    @Override
+    public Object visit(ASTConditionalAndExpression node, Object data) {
+        State state = (State) data;
+        state.booleanOperation(BooleanOp.AND);
+        return super.visit(node, data);
+    }
+
+    @Override
+    public Object visit(ASTConditionalOrExpression node, Object data) {
+        State state = (State) data;
+        state.booleanOperation(BooleanOp.OR);
+        return super.visit(node, data);
+    }
+
+    @Override
+    public Object visit(ASTUnaryExpressionNotPlusMinus node, Object data) {
+        State state = (State) data;
+
+        String op = node.getOperator();
+        if ("!".equals(op)) {
+            state.booleanOperation(null);
+        }
+
+        return super.visit(node, data);
+    }
+
+    @Override
+    public Object visit(ASTBlockStatement node, Object data) {
+        State state = (State) data;
+
+        for (JavaNode child : node.children()) {
+            // This needs to happen because the current 'run' of boolean operations is terminated
+            // once we finish a statement.
+            state.booleanOperation(null);
+
+            child.jjtAccept(this, data);
+        }
+
+        return data;
+    }
+
+    @Override
+    public Object visit(ASTMethodDeclaration node, Object data) {
+        State state = (State) data;
+        state.setMethodName(node.getName());
+        return super.visit(node, data);
+    }
+
+    @Override
+    public Object visit(ASTPrimaryPrefix node, Object data) {
+        State state = (State) data;
+
+        // check if this primary prefix is a method call
+        Iterator<? extends JavaNode> it = node.children().iterator();
+        if (it.hasNext()) {
+            final JavaNode child = it.next();
+            if (child instanceof ASTName) {
+                ASTName name = (ASTName) child;
+                if (name.getNameDeclaration() instanceof MethodNameDeclaration) {
+                    state.methodCall(name.getNameDeclaration().getName());
```

Comment:
I think this is too simple, we should at least check the number of parameters. Or better, check the node that's in the MethodNameDeclaration.

Reply:
Use `getQualifiedName` now - it works, but the method is deprecated. Is it crucial? If so, any hints how to replace this method?